### PR TITLE
lookup parent classes when instance class not found in metadataMap

### PR DESCRIPTION
**Pull request by tobias-trozowski**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description
This PR aims to provide the lookup of parent classes when the instance class is not found in metadataMap when using `ResourceGenerator::fromObject`.

When using doctrine with mezzio-hal i encountered the issue that doctrine proxy classes cannot be mapped and therefore throw an exception.

### CODE REVIEW

**Comment by tobias-trozowski** [comment]
related issue #10

File: test/Assertions.php [hunk 1]
```diff
@@ -26,6 +26,9 @@
 // phpcs:ignore WebimpressCodingStandard.NamingConventions.Trait.Suffix
 trait Assertions
 {
+    /**
+     * @psalm-return ObjectPropertyHydrator::class|ObjectProperty::class
```
Comment (weierophinney) [inline test/Assertions.php:30]:
```suggestion
     * @psalm-return ObjectPropertyHydrator::class&ObjectProperty::class
```

(should be an intersection type)

File: test/ResourceGeneratorTest.php [hunk 2]
```diff
@@ -44,19 +45,28 @@ class ResourceGeneratorTest extends TestCase
 
     use ProphecyTrait;
 
-    /** @var ObjectProphecy|ServerRequestInterface */
+    /**
+     * @var ObjectProphecy|ServerRequestInterface
+     * @psalm-var ObjectProphecy<ServerRequestInterface>
+     */
```
Comment (weierophinney) [inline test/ResourceGeneratorTest.php:51]:
```suggestion
    /** @var ObjectProphecy&ServerRequestInterface */
```

Simpler, and more consistent with what we do in other packages.

File: test/ResourceGeneratorTest.php [hunk 3]
```diff
@@ -44,19 +45,28 @@ class ResourceGeneratorTest extends TestCase
...
     private $request;
 
-    /** @var ObjectProphecy|ContainerInterface */
+    /**
+     * @var ObjectProphecy|ContainerInterface
+     * @psalm-var ObjectProphecy<ContainerInterface>
+     */
```
Comment (weierophinney) [inline test/ResourceGeneratorTest.php:57]:
```suggestion
    /** @var ObjectProphecy&ContainerInterface */
```

File: test/ResourceGeneratorTest.php [hunk 4]
```diff
@@ -44,19 +45,28 @@ class ResourceGeneratorTest extends TestCase
...
     private $hydrators;
 
     /** @var ObjectProphecy|LinkGenerator */
     private $linkGenerator;
 
-    /** @var ObjectProphecy|Metadata\MetadataMap */
+    /**
+     * @var ObjectProphecy|Metadata\MetadataMap
+     * @psalm-var ObjectProphecy<Metadata\MetadataMap>
+     */
```
Comment (weierophinney) [inline test/ResourceGeneratorTest.php:66]:
```suggestion
    /** @var ObjectProphecy&Metadata\MetadataMap */
```

**Review by weierophinney — CHANGES_REQUESTED** [review]
This looks great! My only suggestions are some nit-picks around annotations - use intersection types wherever possible. phpcs no longer complains about them, and Psalm then sees them correctly as well.

File: test/Assertions.php (hunk 1 above)
Comment (weierophinney) [reply test/Assertions.php:30]: Never mind - this is detailing a class string name, so union makes sense.

**Review by weierophinney — APPROVED** [review]
You can ignore each of my suggestions.

When using native mocks, intersection types work perfectly well, and I made the incorrect assumption that this is what was happening here. However, this component is still using Prophecy, and with Prophecy, intersections don't work. The annotations you've provided resolve issues discovered.

**Pull request merged**